File: rene/_trapezoidation/edge.py
from __future__ import annotations

import typing as _t
from typing import Any

import typing_extensions as _te
from reprit.base import generate_repr

from rene import (Orientation,
                  hints as _hints)
from rene._utils import orient


class Edge(_t.Generic[_hints.Scalar]):
    @classmethod
    def from_endpoints(cls,
                       left_point: _hints.Point[_hints.Scalar],
                       right_point: _hints.Point[_hints.Scalar],
                       interior_to_left: bool) -> _te.Self:
        return cls(left_point, right_point, interior_to_left)

    def orientation_of(self,
                       point: _hints.Point[_hints.Scalar]) -> Orientation:
        return orient(self.left_point, self.right_point, point)

    __slots__ = 'interior_to_left', 'left_point', 'right_point'

    def __init__(self,
                 left_point: _hints.Point[_hints.Scalar],
                 right_point: _hints.Point[_hints.Scalar],
                 interior_to_left: bool) -> None:
        assert left_point < right_point, 'Incorrect endpoints order'
        (
            self.interior_to_left, self.left_point, self.right_point
        ) = interior_to_left, left_point, right_point

    def __lt__(self, other: _te.Self) -> Any:
        """Checks if the edge is lower than the other."""
        other_left_orientation = self.orientation_of(other.left_point)
        other_right_orientation = self.orientation_of(other.right_point)
        if other_left_orientation is other_right_orientation:
            return other_left_orientation is Orientation.COUNTERCLOCKWISE
        elif other_left_orientation is Orientation.COLLINEAR:
            return other_right_orientation is Orientation.COUNTERCLOCKWISE
        left_orientation = other.orientation_of(self.left_point)
        right_orientation = other.orientation_of(self.right_point)
        if left_orientation is right_orientation:
            return left_orientation is Orientation.CLOCKWISE
        elif left_orientation is Orientation.COLLINEAR:
            return right_orientation is Orientation.CLOCKWISE
        elif other_right_orientation is Orientation.COLLINEAR:
            return other_left_orientation is Orientation.COUNTERCLOCKWISE
        else:
            return (left_orientation is Orientation.CLOCKWISE
                    if right_orientation is Orientation.COLLINEAR
                    # crossing edges are incomparable
                    else NotImplemented)

    __repr__ = generate_repr(__init__)

File: src/trapezoidation/edge.rs
use std::cmp::Ordering;

use crate::operations::Orient;
use crate::oriented::Orientation;

#[derive(Clone)]
pub(crate) struct Edge<Point> {
    pub(super) left_point: Point,
    pub(super) right_point: Point,
    pub(super) interior_to_left: bool,
}

impl<Point: PartialEq> PartialEq<Self> for Edge<Point> {
    fn eq(&self, other: &Self) -> bool {
        self.left_point.eq(&other.left_point) && self.right_point.eq(&self.right_point) && {
            debug_assert_eq!(self.interior_to_left, other.interior_to_left);
            true
        }
    }
}

impl<Point: PartialEq + Orient> PartialOrd for Edge<Point> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        let other_left_orientation = self.orientation_of(&other.left_point);
        let other_right_orientation = self.orientation_of(&other.right_point);
        if other_left_orientation == other_right_orientation {
            return if other_left_orientation == Orientation::Counterclockwise {
                Some(Ordering::Less)
            } else {
                Some(Ordering::Greater)
            };
        } else if other_left_orientation == Orientation::Collinear {
            return if other_right_orientation == Orientation::Counterclockwise {
                Some(Ordering::Less)
            } else {
                Some(Ordering::Greater)
            };
        }
        let left_orientation = other.orientation_of(&self.left_point);
        let right_orientation = other.orientation_of(&self.right_point);
        if left_orientation == right_orientation {
            if left_orientation == Orientation::Clockwise {
                Some(Ordering::Less)
            } else {
                Some(Ordering::Greater)
            }
        } else if left_orientation == Orientation::Collinear {
            if right_orientation == Orientation::Clockwise {
                Some(Ordering::Less)
            } else {
                Some(Ordering::Greater)
            }
        } else if other_right_orientation == Orientation::Collinear {
            if other_left_orientation == Orientation::Counterclockwise {
                Some(Ordering::Less)
            } else {
                Some(Ordering::Greater)
            }
        } else {
            if right_orientation == Orientation::Collinear {
                if left_orientation == Orientation::Clockwise {
                    Some(Ordering::Less)
                } else {
                    Some(Ordering::Greater)
                }
            } else {
                // crossing edges are incomparable
                None
            }
        }
    }
}

impl<Point: Orient> Edge<Point> {
    pub(super) fn orientation_of(&self, point: &Point) -> Orientation {
        self.left_point.orient(&self.right_point, point)
    }
}
